toml/src/main/kotlin/org/rust/toml/crates/local/CratesLocalIndexUpdater.kt

```diff
@@ -0,0 +1,210 @@
+        val sinceLastUpdate = System.currentTimeMillis() - lastUpdate
+
+        if (sinceLastUpdate < updateIntervalMillis) {
+            if (alarm.isEmpty) {
+                scheduleUpdate(updateIntervalMillis - sinceLastUpdate.toInt())
+            }
+            return
+        }
+        isUpdating = true
+        alarm.cancelAllRequests()
+
+        object : Task.Backgroundable(null, RsTomlBundle.message("rust.update.crates.index.progress.title")) {
+            override fun run(indicator: ProgressIndicator) {
+                LOG.info("crates.io index update started")
+
+                val isSuccessful = updateCratesIoGitIndex(this@CratesLocalIndexUpdater)
+                if (isSuccessful) {
+                    PropertiesComponent.getInstance().setValue(CRATES_IO_INDEX_LAST_UPDATE, System.currentTimeMillis().toString())
+                }
+            }
+
+            override fun onSuccess() {
+                val index = CratesLocalIndexService.getInstance() as? CratesLocalIndexServiceImpl ?: return
+                index.recoverIfNeeded()
+            }
+
+            override fun onFinished() {
+                onUpdateFinished()
+            }
+        }.queue()
+    }
+
+    private fun onUpdateFinished() {
+        LOG.info("crates.io index update finished")
+        isUpdating = false
+        scheduleUpdate(updateIntervalMillis)
+    }
+
+    private fun scheduleUpdate(delay: Int) {
+        if (alarm.isEmpty) {
+            alarm.addRequest(::updateCratesIoGitIndex, delay, true)
+            LOG.info("crates.io index update is scheduled in $delay ms")
+        }
+    }
+
+    override fun dispose() {}
+
+    companion object {
+        private val LOG = logger<CratesLocalIndexUpdater>()
+
+        private const val CRATES_IO_INDEX_LAST_UPDATE = "CRATES_IO_INDEX_LAST_UPDATE"
+
+        private const val DEFAULT_UPDATE_INTERVAL_MIN = 60
+        private const val MIN_UPDATE_INTERVAL_MIN = 1
+
+        private const val UPDATE_PROJECT_DIR_NAME = "crate_index_update_project"
+
+        private val UPDATE_PROJECT_FILES = listOf(
+            // Cargo.toml should contain external dependency to force cargo to load crates index
+            // language=TOML
+            FileTemplate(CargoConstants.MANIFEST_FILE, """
+                [package]
+                name = "crate_index_update_project"
+                version = "0.1.0"
+                edition = "2021"
+
+                # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+                [dependencies]
+                serde = "1.0"
+            """.trimIndent()),
+            FileTemplate("src/lib.rs", "")
+        )
+
+        private val hasOpenRustProject: Boolean
+            get() {
+                // Read action is needed not to get disposed projects
+                // from `com.intellij.openapi.project.ProjectManager.getOpenProjects`
+                return runReadAction {
```

Comment:
It seems like the read action is not needed because you call `hasOpenRustProject` only from EDT where you always have the read lock acquired.